the-state-of-doctests-in-haskell/article.md

```diff
@@ -0,0 +1,253 @@
+there're already a bunch of people using it in real projects.
+The library uses the compiled code - it makes it a lot faster than the alternatives.
+Also, the library doesn't depend on ghc as a library, so it's much more resilient to ghc version changes.
+All of the above makes it more than a viable alternative.
+
+However, the library seems to be closely bound to [`cabal-install`](https://wiki.haskell.org/Cabal): it uses `cabal-install` generated metadata - the `plan.json` file.
+
+Let's consider using `cabal-docspec` to run the doctests in our `doctests-demo` project:
+
+1. First, setup `cabal-install` and the global compiler alongside with all of the auxiliary tooling.
+We suggest using [ghcup](https://www.haskell.org/ghcup/) to achieve this.
+
+2. Build the project using `cabal`:
+
+    `cabal v2-build`
+
+3. Download `cabal-docspec` binaries from [the releases page](https://github.com/phadej/cabal-extras/releases).
+You can also consider building the library from source, following the instructions [here](https://github.com/phadej/cabal-extras).
+
+    ```
+    curl -sL https://github.com/phadej/cabal-extras/releases/download/cabal-docspec-0.0.0.20211114/cabal-docspec-0.0.0.20211114.xz > cabal-docspec.xz
+    xz -d < cabal-docspec.xz > "$HOME"/.local/bin/cabal-docspec
+    rm -f cabal-docspec.xz
+    chmod a+x "$HOME"/.local/bin/cabal-docspec
+    ```
+
+4. Now, run the doctest examples using `cabal-docspec`:
+
+    `cabal-docspec`
+
+5. The command above should fail with the following error:
+
+    ```
+    expected: "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
+    but got:
+              ^
+              <interactive>:10:1: error:
+                  Variable not in scope: encode :: Anime -> t
+    ```
+
+6. The above happens because the library handles things a little differently - it requires modules to be explicitly imported/exported.
+Let's edit our doctests a little to make `cabal-docspec` work:
+
+    ```haskell
+    -- src/Sample.hs
+
+    {-# LANGUAGE OverloadedStrings #-}
+    {-# LANGUAGE TemplateHaskell   #-}
+
+    module Sample where
+
+    import Data.Text
+    import Data.Aeson
+    import Data.Aeson.TH
+
+    data Anime =
+      Anime { title  :: Text
+            , rating :: Double
+            }
+
+    $(deriveJSON defaultOptions ''Anime)
+
+    -- |
+    -- >>> import Data.Aeson
+    -- >>> encode favourite
+    -- "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
+    --
+    favourite :: Anime
+    favourite = Anime "One-Punch Man" 8.9
+    ```
+
+7. Now, the `cabal-docspec` command should succeed:
+
+      ```
+      Total:         2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
+      Examples:      2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
+      ```
+
+Thanks for reading!
+Subscribe to our newsletter to stay tuned - `cabal-docspec` definitely needs a dedicated post covering a little more about its configuration and internals!
```

Comment:
Wouldn't mind if the info was here.

Reply:
Cool. Now we just can't forget to actually create the new post in the future. ;)